lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -329,6 +405,29 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
                                          status);
     return (status == 0);
   }
+
+  // Reads a fd and executes callback on each read 
+  // until either EOF is encountered or read() returns 
+  // with an error.
+  // Returns true if EOF is found.
+  // Returns false if read() returns negative value.
```

Comment:
If this function handles the errors, then it can return a bool just based on whether the fd should be closed, which I think is a simpler API.

Also, note that it is risky to depend on `errno` in the way this code does, because `errno` can change just about anywhere (anything that might trigger a syscall).